### DISCUSSION TITLE
Add option to import .csl File through URL

### DISCUSSION THREAD

**Discussion by philipprhrd** [zepinglee/citeproc-lua#69] [Ideas]
Hi,
I want to create a template for LaTeX using the citation-style-language package. Currently the only option for importing a citation style is to store the .csl file locally in the project. But what about importing the style from a url that points to a .csl file?

`\usepackage[style = {https://github.com/..../style.csl}]{citation-style-language}`

I think this would be a great idea for keeping a citation style up to date for different projects at the same time.

**Comment by zepinglee** [comment]
That's a reasonable feature and technically it can be implemented with the `luasocket` module in LuaTeX. Let me have a try.

**Reply by zepinglee** [reply]
Unfortunately I can't fetch the CSL style from GitHub with `luasocket`.

```lua
local http = require("socket.http")
local ltn12 = require("ltn12")

local resp = {}
local body, code, header = http.request({
  -- url = "http://www.zotero.org/styles/american-physics-society",
  -- url = "https://github.com/citation-style-language/styles/raw/v1.0.2/american-physics-society.csl",
  url = "https://raw.githubusercontent.com/citation-style-language/styles/v1.0.2/american-physics-society.csl",
  method = "GET",
  redirect = true,
  sink = ltn12.sink.table(resp),
})

print(body, code, header)

for key, value in pairs(header) do
  print(key, value)
end

print(table.concat(resp))
```

```bash
$ texlua test.lua
nil     connection refused      nil
test.lua:16: bad argument #1 to 'for iterator' (table expected, got nil)
```

**Reply by philipprhrd** [reply]
Hi, thank you for your answer. Is it a lua thing or github thing?
My personal use case would be from a gitlab self hosted instance. 

Thank you for testing it out!

**Reply by zepinglee** [reply]
> Is it a lua thing or github thing?

It should be related to he Lua library. The Python library `requests` works with those three URLs but LuaTeX doesn't provide a convenient library like that.
